### PR TITLE
[Reland] Refactor caching device allocator utils

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129919
* __->__ #130923

# Motivation
Following [[RFC] Intel GPU Runtime Upstreaming for Allocator ](https://github.com/pytorch/pytorch/issues/116322), this PR aims to refactor caching device allocator utils to improve code reuse usage.
This is the first PR, we could prepare some follow-up PRs continuing to refactor the device caching allocator.

### CODE REVIEW

File: c10/core/CachingDeviceAllocator.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  Stat oversize_allocations;
+
+  // COUNT: total number of oversize blocks requiring malloc
+  Stat oversize_segments;
+
+  // COUNT: total number of synchronize_and_free_events() calls
+  int64_t num_sync_all_streams = 0;
+
+  // COUNT: total number of device memory allocation calls. This includes both
+  // cuMemMap and cudaMalloc.
```
Comment (EikanWang) [inline c10/core/CachingDeviceAllocator.h:100]:
```suggestion
  // mapped and malloced memory
```

File: c10/core/CachingDeviceAllocator.h [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  int64_t num_device_alloc = 0;
+
+  // COUNT: total number of device memory deallocation calls. This includes both
+  // cuMemUnmap and cudaFree.
```
Comment (EikanWang) [inline c10/core/CachingDeviceAllocator.h:104]:
```suggestion
  //un-mapped and free memory.
```

**Review by EikanWang — APPROVED** [review]

**Review by gujinghui — APPROVED** [review]

**Review by albanD — APPROVED** [review]
Thanks!

**Review by eqy — APPROVED** [review]

**Comment by ZainRizvi** [comment]
@pytorchbot revert -c ghfirst -m "Sorry but this appears to be causing internal tests to fail with errors like `error: no type named 'DeviceStats' in namespace 'xxx::xxx:xxxAllocator'; did you mean 'DeviceStatus'?`"

@albanD, can you please help get this PR re-merged?  Diff with failures is D61897820

File: c10/cuda/CUDACachingAllocator.h [hunk 3]
```diff
@@ -50,73 +50,6 @@ namespace c10::cuda::CUDACachingAllocator {
...
-  AGGREGATE = 0,
-  SMALL_POOL = 1,
-  LARGE_POOL = 2,
-  NUM_TYPES = 3 // remember to update this whenever a new stat type is added
-};
-
-typedef std::array<Stat, static_cast<size_t>(StatType::NUM_TYPES)> StatArray;
-
-// Struct containing memory allocator summary statistics for a device.
-struct DeviceStats {
```
Comment (albanD) [inline c10/cuda/CUDACachingAllocator.h:70]: This is the entry that some internal system looks for.
Can you re-expose the new version here to preserve the public API that was exposed from this header?

Comment (guangyey) [reply c10/cuda/CUDACachingAllocator.h:70]: I expose `c10::CachingDeviceAllocator::DeviceStats` as `c10::cuda::CUDACachingAllocator::DeviceStats` for BC. Could I capture your point?

**Review by albanD — APPROVED** [review]
This is tricky to import but from looking at the code, this was the only used API being removed. Let's merge again as is!

All the dataloder failures are from trunk btw

**Comment by guangyey** [comment]
Thanks, try to reland.

**Comment by kit1980** [comment]
@guangyey you probably should work with someone from Meta to land this, I see "Relocation overflow has occured" but I don't understand what's going on. The diff number is D62215424

**Comment by guangyey** [comment]
> @guangyey you probably should work with someone from Meta to land this, I see "Relocation overflow has occured" but I don't understand what's going on. The diff number is D62215424

Thanks @kit1980, I would ask for help from @albanD

**Comment by albanD** [comment]
@pytorchbot merge

From 1-1 chat, internal failures are not related and can be ignred.

**Comment by albanD** [comment]
@guangyey can you rebase and resolve merge conflicts for this?

**Comment by guangyey** [comment]
> @guangyey can you rebase and resolve merge conflicts for this?

Sure, Thanks very much~

**Comment by guangyey** [comment]
It seems to be unrelated failures, try to reland again.
@pytorchbot merge -i

**Pull request closed without merging**